Add `-Z hint-mostly-unused` to tell rustc that most of a crate will go unused

joshtriplett opened the pull request:
This hint allows the compiler to optimize its operation based on this assumption, in order to compile faster. This is a hint, and does not guarantee any particular behavior.

This option can substantially speed up compilation if applied to a large dependency where the majority of the dependency does not get used. This flag may slow down compilation in other cases.

Currently, this option makes the compiler defer as much code generation as possible from functions in the crate, until later crates invoke those functions. Functions that never get invoked will never have code generated for them. For instance, if a crate provides thousands of functions, but only a few of them will get called, this flag will result in the compiler only doing code generation for the called functions. (This uses the same mechanisms as cross-crate inlining of functions.) This does not affect `extern` functions, or functions marked as `#[inline(never)]`.

This option has already existed in nightly as `-Zcross-crate-inline-threshold=always` for some time, and has gotten testing in that form. However, this option is still unstable, to give an opportunity for wider testing in this form.

Some performance numbers, based on a crate with many dependencies having just *one* large dependency set to `-Z hint-mostly-unused` (using Cargo's `profile-rustflags` option):

A release build went from 4m07s to 2m04s.
    
A non-release build went from 2m26s to 1m28s.

joshtriplett committed:
Add `-Z hint-mostly-unused` to tell rustc that most of a crate will go unused

This hint allows the compiler to optimize its operation based on this
assumption, in order to compile faster. This is a hint, and does not
guarantee any particular behavior.

This option can substantially speed up compilation if applied to a large
dependency where the majority of the dependency does not get used. This flag
may slow down compilation in other cases.

Currently, this option makes the compiler defer as much code generation as
possible from functions in the crate, until later crates invoke those
functions. Functions that never get invoked will never have code generated for
them. For instance, if a crate provides thousands of functions, but only a few
of them will get called, this flag will result in the compiler only doing code
generation for the called functions. (This uses the same mechanisms as
cross-crate inlining of functions.) This does not affect `extern` functions, or
functions marked as `#[inline(never)]`.

Some performance numbers, based on a crate with many dependencies having
just *one* large dependency set to `-Z hint-mostly-unused` (using
Cargo's `profile-rustflags` option):

A release build went from 4m07s to 2m04s.

A non-release build went from 2m26s to 1m28s.

GuillaumeGomez commented:
r? compiler

workingjubilee commented:
This should be added as an unstable option first?

joshtriplett commented:
@workingjubilee This option has already existed in nightly as `-Zcross-crate-inline-threshold=always` for some time, and has gotten testing in that form.

The only difference being made between that option and this one is that this one respects `#[inline(never)]`.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12837157842/job/35800269136) [(plain)](https://github.com/rust-lang/rust/commit/4d1b2cb8966abbf1fc9be773c5fff6cf0453d6d3/checks/35800269136/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 28.4s done
#21 DONE 34.4s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.99s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
```

</details>

workingjubilee commented:
```

     tests::test_codegen_options_tracking_hash ... FAILED
  
  
  failures:
  
  ---- tests::test_codegen_options_tracking_hash stdout ----
  
  thread 'tests::test_codegen_options_tracking_hash' panicked at compiler/rustc_interface/src/tests.rs:618:5:
  assertion `left != right` failed
    left: false
   right: false
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
  
```

uhm

Urgau commented on compiler/rustc_interface/src/tests.rs:
```diff
@@ -615,6 +615,7 @@ fn test_codegen_options_tracking_hash() {
     tracked!(embed_bitcode, false);
     tracked!(force_frame_pointers, FramePointer::Always);
     tracked!(force_unwind_tables, Some(true));
+    tracked!(hint_mostly_unused, false);
```
It must be different from the default value of the option.
```suggestion
    tracked!(hint_mostly_unused, true);
```

lqd replied:
```suggestion
    tracked!(hint_mostly_unused, true);
```

This s should have a non default value, to ensure options are correctly tracked when changed.

workingjubilee commented:
@joshtriplett:
> @workingjubilee This option has already existed in nightly as -Zcross-crate-inline-threshold=always for some time, and has gotten testing in that form.
>
> The only difference being made between that option and this one is that this one respects `#[inline(never)]`.

I see. I suppose that makes sense, though it brings these thoughts to mind:
- In general, T-compiler has a much lower threshold for unstable flags that are not intended for serious or permanent usage downstream. They're often added as debugging aids. I am not sure if this was intended for public usage when it was originally added.
- All inlining annotations are hints, and this is also noted as a hint, yet it changes the behavior of another inlining hint, and in that way deviates from its unstable variation.
- Notably, `inline(never)` is often relied on to some degree by programmers as working.
- This describes itself as a hint, but also is motivated on the basis that it will be very impactful, which suggests its behavior will be immediately relied on. The combination of relying-on-inlining-but-also-relying-on-not-inlining makes me wonder if it is definitely correct? There are a lot of weird nuances to our inlining rules.

As a result, I am curious if the rest of T-compiler believes changing the behavior with respect to `inline(never)` is trivial.

Perhaps this should land with a test that demonstrates that it respects `#[inline(never)]` correctly, now?

workingjubilee commented:
( I'm also kinda curious if we should adjust how we handle cross-crate inlining by default, tbh. )

lqd commented:
You may have missed it Jubilee, so hop on to https://rust-lang.zulipchat.com/#narrow/stream/233931-xxx/topic/Add.20.60-C.20hint-mostly-unused.60.20option.20compiler-team.23829 to ask your cool questions and comments :)

Urgau commented on src/doc/rustc/src/codegen-options/index.md:
```diff
@@ -172,6 +172,25 @@ values:
 
 The default if not specified depends on the target.
 
+## hint-mostly-unused
```
Given that the flag is now unstable/nightly-gated, this documentation should be moved to the unstable-book (under `src/doc/unstable-book/src/compiler-flags`) and a tracking issue should be created.

lqd replied:
No need for a tracking issue before consensus tho right ? This still needs fcp anyways. (And given the reservations in the thread some more tests and all will be needed before stabilization.)

Urgau replied:
Sure, the creation of the tracking issue can wait until the MCP is accepted and the PR nearing approval.

joshtriplett replied:
I've updated this to be an unstable option and documented in the unstable book.

apiraino commented:
Discussed in T-compiler triage [on Zulip](https://rust-lang.zulipchat.com/#narrow/channel/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202025-01-23.5D/near/495530702).

[MCP#829](https://github.com/rust-lang/compiler-team/issues/829) still needs to be seconded and there seems to be still a few open design questions. Also, do we have some numbers about the real world impact (both positive and negative)?. Maybe it makes sense to nominate again once design and scope are clearer.

By the way @saethlin do you think you have some capacity to dedicate here?

@rustbot label -I-compiler-nominated

saethlin commented on compiler/rustc_mir_transform/src/cross_crate_inline.rs:
```diff
@@ -50,6 +50,12 @@ fn cross_crate_inlinable(tcx: TyCtxt<'_>, def_id: LocalDefId) -> bool {
         _ => {}
     }
 
+    // If the crate is likely to be mostly unused, use cross-crate inlining to defer codegen until
+    // the function is referenced, in order to skip codegen for unused functions.
+    if tcx.sess.opts.cg.hint_mostly_unused {
+        return true;
+    }
+
```
Placing this logic below the check for `InlineAttr::Never` does not change whether a call to a `#[inline(never)]` function can be inlined, it changes when and where `#[inline(never)]` functions get compiled. With the logic down here, `#[inline(never)]` functions and their callees (recursively) are not delayed.

The exact motivation for putting this logic here instead of above the check for `codegen_fn_attrs.inline` should be clearly recorded. I don't rightly remember what it was, but for example panic backtrace trimming still works even if it is moved.

joshtriplett replied:
Making sure inline-never functions don't get affected was an intentional goal here. I'll add a comment.

joshtriplett replied:
I've added a comment and documentation for this.

saethlin commented on src/doc/rustc/src/codegen-options/index.md:
```diff
@@ -172,6 +172,25 @@ values:
+The compiler can optimize its operation based on this assumption, in order to
+compile faster. This is a hint, and does not guarantee any particular behavior.
+
+This option can substantially speed up compilation if applied to a large
+dependency where the majority of the dependency does not get used. This flag
```
Can we provide any indication of how to do this? I've suggested some people try adding `-Zcross-crate-inline-threshold=always` to just one dependency and nobody seems to know how to do this offhand. If we're going to suggest people do a thing that's poorly understood I'd like to at least have a link to the Cargo docs about profile overrides.

joshtriplett replied:
Sure, will do.

joshtriplett replied:
I added documentation explaining how to do this with Cargo.

joshtriplett commented:
Based on discussion about this at RustWeek, it sounded like this proposal had given the impression of wanting to rush stabilization, rather than being an unstable option to test and gather experience reports. Based on that discussion, I've reworked this PR and opened [a new MCP](https://github.com/rust-lang/compiler-team/issues/883) to make it explicitly clear that this is an unstable option intended for a call for testing and for plumbing as a nightly Cargo option.

I've updated the implementation of this to be a `-Z` option, and added more documentation for it in the unstable book (including how to use Cargo to selectively enable it for specific crates).

joshtriplett commented:
@rustbot ready

saethlin commented:
@bors r+

bors commented:
:pushpin: Commit 1b23b64be4bf9e827cbc9d7c4012da77e8f05a85 has been approved by `saethlin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=saethlin 1b23b64be4bf9e827cbc9d7c4012da77e8f05a85 -->
<!-- homu: {"type":"Approved","sha":"1b23b64be4bf9e827cbc9d7c4012da77e8f05a85","approver":"saethlin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
